### PR TITLE
248. Strobogrammatic Number III

### PR DESCRIPTION
**Pull request by altayhunoglu**
Resolves: https://github.com/spiralgo/algorithms/issues/322

Even if the question is not in Curated 170, it is important as it will be a basis for: https://github.com/spiralgo/algorithms/issues/228

We use almost the same code with our solution in: https://github.com/spiralgo/algorithms/pull/305

This time: 
- We need to control if the number is in the `range` given with `low` and `high`.
- We will just count the numbers instead of adding them to a list.

### CODE REVIEW

File: src/main/java/algorithms/hard/StrobogrammaticNumber3.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        for (char[] pair : reflectives) {
+            char ch = pair[0];
+            numBuilder[left] = ch;
+            numBuilder[right] = pair[1];
+            if (numBuilder.length != 1 && numBuilder[0] == '0') {
+                continue;
+            } else if (left == right && (ch == '6' || ch == '9')) {
+                continue;
+            }
```
Comment (altayhunoglu) [inline src/main/java/algorithms/hard/StrobogrammaticNumber3.java:50]:
```suggestion
   if ((numBuilder.length != 1 && numBuilder[0] == '0')  || (left == right && (ch == '6' || ch == '9'))) {
            continue;
           }
```

File: src/main/java/algorithms/hard/StrobogrammaticNumber3.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            numBuilder[right] = pair[1];
+   if ((numBuilder.length != 1 && numBuilder[0] == '0')  || (left == right && (ch == '6' || ch == '9'))) {
+            continue;
+           }
+            count += createNum(low, high, numBuilder, left + 1, right - 1);
+        }
+
+        return count;
+
+    }
```
Comment (ErdemT09) [inline src/main/java/algorithms/hard/StrobogrammaticNumber3.java:54]: We can extract the length of the strings to fields to not call their methods again.

```suggestion
    int highLen;
    int lowLen;
    int currLen;

    public int strobogrammaticInRange(String low, String high) {

        if (low == null || high == null || low.length() > high.length()
                || (low.length() == high.length() && low.compareTo(high) > 0)) {
            return 0;
        }

        highLen = high.length();
        lowLen = low.length();

        int count = 0;
        for (int len = lowLen; len <= highLen; len++) {
            currLen = len;
            count += createNum(low, high, new char[len], 0, len - 1);
        }
        return count;

    }

    int createNum(String low, String high, char[] numBuilder, int left, int right) {
        if (left > right) {

            String s = new String(numBuilder);
            if ((currLen == lowLen && s.compareTo(low) < 0) || (currLen == highLen && s.compareTo(high) > 0)) {
                return 0;
            } else {
                return 1;
            }
        }

        int count = 0;

        for (char[] pair : reflectives) {
            char ch = pair[0];
            numBuilder[left] = ch;
            numBuilder[right] = pair[1];
            if ((currLen != 1 && numBuilder[0] == '0') || (left == right && (ch == '6' || ch == '9'))) {
                continue;
            }
            count += createNum(low, high, numBuilder, left + 1, right - 1);
        }

        return count;

    }
```

**Comment by ErdemT09** [comment]
We can actually just convert the strings to character arrays and compare the arrays themselves, which would have been faster, but it's sometimes better to use what's built-in than trying to reinvent a better wheel.

**Comment by altayhunoglu** [comment]
> Was einfacher für einen Mensch zu coden ist, ist manchmal besser als, was für eine Maschine schneller ist.
  
 You are right and thanks for the German translation.
 Mein guter kamerad. 
 It helps a lot.

**Review by ErdemT09 — APPROVED** [review]
Lean code following up from the prequel problem. Danke wieder für die Mitarbeit und Beitragleistung hier. Approved

**Pull request merged**